src/Common/src/Microsoft/Win32/SafeHandles/NtlmSecuritySafeHandles.cs

```diff
@@ -0,0 +1,271 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using System.Security.Authentication;
+using System.Security.Authentication.ExtendedProtection;
+using System.Security.Cryptography;
+using System.Security.Cryptography.X509Certificates;
+using System.Text;
+
+namespace Microsoft.Win32.SafeHandles
+{
+    /// <summary>
+    /// Wrapper around a ntlm_buf*
+    /// </summary>
+    internal sealed class SafeNtlmBufferHandle : SafeHandle
+    {
+        public SafeNtlmBufferHandle()
+           : base(IntPtr.Zero, true)
+        {
+        }
+
+        public byte[] ToByteArray(int length, int offset)
+        {
+            Debug.Assert(length >= 0, "negative length of buffer");
+            Debug.Assert(offset < length, "invalid offset for  buffer");
+            byte[] target = new byte[length];
+            Interop.NetSecurity.ExtractNtlmBuffer(this, target, capacity: length, offset: offset);
+            return target;
+        }
+
+        public override bool IsInvalid
+        {
+            get { return handle == IntPtr.Zero; }
+        }
+
+        protected override bool ReleaseHandle()
+        {
+            Interop.NetSecurity.HeimNtlmFreeBuf(handle);
+            SetHandle(IntPtr.Zero);
+            return true;
+        }
+    }
+
+    /// <summary>
+    /// Wrapper around a session key used for signing
+    /// </summary>
+    internal sealed class SafeNtlmKeyHandle : SafeHandle
+    {
+        private uint _sequenceNumber;
+        private readonly bool _isSealingKey;
+        private readonly byte[] _digest;
+        private SafeEvpCipherCtxHandle _cipherContext;
+
+        // From MS_NLMP SIGNKEY at https://msdn.microsoft.com/en-us/library/cc236711.aspx
+        private const string s_keyMagic = "session key to {0}-to-{1} {2} key magic constant\0";
+        private const string s_client = "client";
+        private const string s_server = "server";
+        private const string s_signing = "signing";
+        private const string s_sealing = "sealing";
+
+        public SafeNtlmKeyHandle(byte[] key, bool isClient, bool isSealingKey)
+            : base(IntPtr.Zero, true)
+        {
+            string keyMagic = string.Format(s_keyMagic, isClient ? s_client : s_server,
```

Comment:
Rather than calling string.Format every time we make one of these handles this way, can you just make the four constants be, well, constants?  That'd save both space and time.